### DISCUSSION TITLE
Enforce `ISerializable` for data types

### DISCUSSION THREAD

**Discussion by longfin** [planetarium/libplanet#2450] [Ideas]
When an object is serialized in .NET, it is often necessary for that field to be serializable. (e.g., best practices about custom exceptions), And when the object being serialized is out of our control, this can be a big problem.

So rather, I suggest that we proactively implement [`ISerialzable`][ISerialzable] for the types we expect to be treated as data in Libplanet, and prevent this implicit field serialization. This can be easily applied to types for which we already have (de)serialzation methods, such as `Block<T>` or `Transaction<T>`. 

The following is an example when applied to `Transaction<T>`.

```csharp
 public sealed class Transaction<T> : IEquatable<Transaction<T>>, ITxExcerpt, ISerializable
        where T : IAction, new()
{
        protected Transaction<T>(SerializationInfo info, StreamingContext context)
            : this(new Bencodex.Codec().Decode(info?.GetValue<byte[]>("asBencodex")))
        {
        }

        public void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            info.AddValue("asBencodex", new Bencodex.Codec().Encode(ToBencodex()));
        }
}
```

[ISerialzable]: https://learn.microsoft.com/en-us/dotnet/api/system.runtime.serialization.iserializable

**Comment by dahlia** [comment]
It would be great if we can enforce it using static analysis (lint).  I have no idea if there are any lints having such rule though…

**Comment by greymistcube** [comment]
I wish it was that simple, but I think we might want to also enforce something like `IBencodexSerializable` interface for data types (although this is for a slightly different reason; hence me "pre-emptively" working on #2457). In any case, byte array serialization and `Bencodex` serialization is left freely up to each `class` at the moment. This has became an issue while working on PBFT as PBFT will tack on many more data types to existing one, and there is no real standard to follow. 🙄

If we have a strong enough contract, then I think it'd help for outside/new contributors to follow more easily. 😗

**Comment by longfin** [comment]
On the other hand, this effort may be unnecessary or will affect only a very limited duration since .NET seems to decide to step down the legacy serialization API, including `ISerializable`. ([see also](https://github.com/dotnet/designs/blob/8b779c3ebae061db7f4d546153be5796ebcfce7c/accepted/2020/better-obsoletion/binaryformatter-obsoletion.md#new-obsoletions-in-net-8))
